### PR TITLE
service/cfg: Separate out config block defaults and create when not found.

### PR DESCRIPTION
**Pull request by Steveice10**
Separates out all of the default config block definitions to a separate source unit. Instead of FormatConfig creating each individual block in code one-by-one, we now have a well-defined map of block IDs to their access flags and default initial values.

This allows us to extend config block creation into GetConfigBlockPointer calls. When a request is made for a config block that does not exist, it now checks to see if a default is defined. If so, the block will be created on the fly, otherwise we will log an error and return an error code.

This is to account for the ever evolving config block defaults. If a user already has a config savegame from a previous FormatConfig operation, and the version of Citra that created it did not define a specific block that is now known, we can create that block and add it to the savegame on the fly.

This should fix issues people may have when using the home menu with a config file made by older versions of Citra without certain blocks defined. It also adds in a missing default for block 0xC0002 (parental restrictions email address) which is needed by the home menu.

I've tested this in the home menu with:
* No config savegame defined, to have freshly formatted data.
* Block 0xC0002 missing, to test adding new blocks on the fly at read time.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6716)

### CODE REVIEW

File: src/core/hle/service/cfg/cfg_defaults.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include "core/hle/service/cfg/cfg.h"
+
+namespace Service::CFG {
+
+/// The default parameters of a config block.
+struct ConfigBlockDefaults {
+    AccessFlag access_flags;
+    std::span<const u8> data;
+
+    ConfigBlockDefaults(AccessFlag _access_flags, const void* _data, std::size_t _size)
```
Comment (SachinVin) [inline src/core/hle/service/cfg/cfg_defaults.h:19]: Nit, we generally have underscores at the end

File: src/core/hle/service/cfg/cfg_defaults.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#include "core/hle/service/cfg/cfg.h"
+#include "core/hle/service/cfg/cfg_defaults.h"
+
+namespace Service::CFG {
+
+constexpr std::size_t LOCATION_NAME_SIZE = 0x40;
+std::array<u16_le, LOCATION_NAME_SIZE> MakeLocationName(std::string_view input) {
+    auto converted = Common::UTF8ToUTF16(input);
+    std::array<u16_le, LOCATION_NAME_SIZE> out{};
+    for (std::size_t i = 0; i < 16; ++i) {
```
Comment (SachinVin) [inline src/core/hle/service/cfg/cfg_defaults.cpp:16]: Nit magic number, but also any reason we can't use `u"United States"`?

File: src/core/hle/service/cfg/cfg_defaults.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+constexpr u8 WASHINGTON_DC_STATE_ID = 2;
+static const std::array<u16_le, LOCATION_NAME_SIZE> UNITED_STATES_COUNTRY_NAME =
+    MakeLocationName("United States");
+static const std::array<u16_le, LOCATION_NAME_SIZE> WASHINGTON_DC_STATE_NAME =
+    MakeLocationName("Washington DC");
+
+constexpr u64_le DEFAULT_USER_TIME_OFFSET = 0;
+constexpr BacklightControls DEFAULT_BACKLIGHT_CONTROLS{0, 2};
+/**
+ * TODO(Subv): Find out what this actually is, these values fix some NaN uniforms in some games,
```
Comment (SachinVin) [inline src/core/hle/service/cfg/cfg_defaults.cpp:32]: Is this todo still relevant?

Comment (Steveice10) [reply src/core/hle/service/cfg/cfg_defaults.cpp:32]: I think so, we still don't really know exactly what the values are. Either way I didn't make any changes to this constant so I'd rather just leave the note on it as-is.

File: src/core/hle/service/cfg/cfg_defaults.cpp (hunk 2 above)
Comment (Steveice10) [reply src/core/hle/service/cfg/cfg_defaults.cpp:16]: Figured out a cleaner way to do it based on this comment, thanks.

File: src/core/hle/service/cfg/cfg.h [hunk 4]
```diff
@@ -51,7 +119,49 @@ enum SoundOutputMode { SOUND_MONO = 0, SOUND_STEREO = 1, SOUND_SURROUND = 2 };
 struct EULAVersion {
     u8 minor;
     u8 major;
+    INSERT_PADDING_BYTES(2);
+};
+static_assert(sizeof(EULAVersion) == 4, "EULAVersion must be exactly 0x4 bytes");
+
+struct UsernameBlock {
+    char16_t username[10]; ///< Exactly 20 bytes long, padded with zeros at the end if necessary
```
Comment (SachinVin) [inline src/core/hle/service/cfg/cfg.h:127]: std::array

File: src/core/hle/service/cfg/cfg_defaults.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+bool HasDefaultConfigBlock(ConfigBlockID block_id) {
+    return DEFAULT_CONFIG_BLOCKS.contains(block_id);
+}
+
+const ConfigBlockDefaults& GetDefaultConfigBlock(ConfigBlockID block_id) {
+    return DEFAULT_CONFIG_BLOCKS.at(block_id);
+}
+
+} // namespace Service::CFG
```
Comment (SachinVin) [inline src/core/hle/service/cfg/cfg_defaults.cpp:133]:
```suggestion
} // namespace Service::CFG

```

File: src/core/hle/service/cfg/cfg.h (hunk 4 above)
Comment (Steveice10) [reply src/core/hle/service/cfg/cfg.h:127]: This is code moved from the existing definitions in the cpp file, but I can update this.

**Review by SachinVin — APPROVED** [review]

File: src/core/hle/service/cfg/cfg.h [hunk 6]
```diff
@@ -51,7 +119,50 @@ enum SoundOutputMode { SOUND_MONO = 0, SOUND_STEREO = 1, SOUND_SURROUND = 2 };
...
+
+struct BirthdayBlock {
+    u8 month; ///< The month of the birthday
+    u8 day;   ///< The day of the birthday
+};
+static_assert(sizeof(BirthdayBlock) == 2, "BirthdayBlock must be exactly 2 bytes");
+
+struct ConsoleModelInfo {
+    u8 model;      ///< The console model (3DS, 2DS, etc)
+    u8 unknown[3]; ///< Unknown data
```
Comment (FearlessTobi) [inline src/core/hle/service/cfg/cfg.h:142]: Can these be INSERT_PADDING_BYTES or at least an std::array?

**Review by FearlessTobi — APPROVED** [review]

**Pull request merged**